doc: net: Add socket service API usage documentation

jukkar opened the pull request:
Add information how to use socket service API from application point of view.

Fixes #86982

jukkar committed:
doc: net: Add socket service API usage documentation

Add information how to use socket service API from application
point of view.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
@pdgendt I think I directed you to wrong example of the socket service API. The `echo_service` sample is probably the easiest app to understand what is going on with the API.

pdgendt commented on doc/connectivity/networking/api/socket_service.rst:
```diff
@@ -0,0 +1,142 @@
+Overview
+********
+
+The socket service API can be used to install a handler that is called if there
+is data received to a socket. The API helps to avoid creating a dedicated thread
+for each TCP or UDP services that the application provides. Instead one thread
+is created which can serve data to multiple listening sockets which saves memory
+as only one thread needs to be created in the system in this case.
```
```suggestion
The socket service API can be used to install a handler that is called if there
is data received on a socket. The API helps to avoid creating a dedicated thread
for each TCP or UDP service that the application provides. Instead one thread
is created that serves data to multiple listening sockets which saves memory
as only one thread needs to be created in the system in this case.
```

pdgendt commented on doc/connectivity/networking/api/socket_service.rst:
```diff
@@ -0,0 +1,142 @@
+
+* :c:type:`net_socket_service_handler_t`
+
+  User specified callback which is called when data is received to the
+  listening socket.
```
```suggestion
  User specified callback which is called when data is received on the
  listening socket.
```

pdgendt commented on doc/connectivity/networking/api/socket_service.rst:
```diff
@@ -0,0 +1,142 @@
+Application Overview
+********************
+
+When socket service API is enabled, the application must create the service like
+this:
```
```suggestion
If the socket service API is enabled, an application must create the service like
this:
```

pdgendt commented on doc/connectivity/networking/api/socket_service.rst:
```diff
@@ -0,0 +1,142 @@
+  User specified callback which is called when data is received to the
+  listening socket.
+
+Application Overview
```
The overview really only shows a "simple" UDP service, while it is more complex for TCP as it creates new socket file descriptors for connecting clients.
It would be nice to have that distinction in the documentation.

jukkar replied:
Yes, I thought about this, but did not want to write complete application as the `echo-service` application has that implemented. But we could put something more about TCP and what to do with the accepted socket.

pdgendt replied:
Yes, one thing that tricked me, and would be nice to have documented, is that the `accept` call can be done in the callback too. This isn't the case in the `echo-service` sample, where instead a loop is used for a single client.

In most cases, I think, services should serve multiple clients.

jukkar replied:
Sure, we could enhance the echo-service sample to be a bit more realistic. Patches are welcome to enhance the sample.

pdgendt commented on doc/connectivity/networking/api/socket_service.rst:
```diff
@@ -0,0 +1,142 @@
+.. _socket_service_interface:
+
+Socket Service
+##############
```
Feels like the plural is better suited in the title?

```suggestion
Socket Services
###############
```

pdgendt reviewed (commented):
Thanks @jukkar for this, a nice starting point. A few minor remarks, and one "bigger" remark concerning TCP clients.

jukkar commented:
The generated documentation looks like this
https://builds.zephyrproject.io/zephyr/pr/87110/docs/connectivity/networking/api/socket_service.html

jukkar committed:
samples: net: echo_service: Fix documentation

Clarify the documentation in the README.rst file. Also remove the
obsolete comment from main.c as it is no longer correct.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
* Updated according to comments
* Fixed the documentation in the echo-service sample

jukkar commented on doc/connectivity/networking/api/socket_service.rst:
```diff
@@ -0,0 +1,219 @@
+   sockfd_udp[0].events = POLLIN;
+
+   /* Register UDP socket to service handler */
+   ret = net_socket_service_register(&service_udp, sockfd_udp,
+				     ARRAY_SIZE(sockfd_udp), NULL);
+   if (ret < 0) {
+	LOG_ERR("Cannot register socket service handler (%d)", ret);
+	return ret;
+   }
+
+   /* Application logic happens here. When application is ready to
+    * quit, one should unregister the socket service and close the
+    * socket.
+    */
+
+   (void)net_socket_service_unregister(&service_udp);
```
As the example was taken from echo_service sample, I will use its variable names so the correct one is `service_udp`. I will send a fix for this.
